lsp-mode.el

```diff
@@ -1859,42 +1852,88 @@ WORKSPACE is the workspace that contains the progress token."
+
 (defun lsp--diagnostics-modeline-statistics ()
   "Calculate diagnostics statistics based on `lsp-diagnostics-modeline-scope'"
   (let ((diagnostics (cond
-                      ((equal :file lsp-diagnostics-modeline-scope)
-                       (lsp--get-buffer-diagnostics))
-                      (t (->> (eq :workspace lsp-diagnostics-modeline-scope)
-                              (lsp-diagnostics)
-                              (ht-values)
-                              (-flatten))))))
-    (->> diagnostics
-         (-group-by #'lsp:diagnostic-severity?)
-         (-sort (-lambda ((left) (right))
-                  (> right left)))
-         (-map (-juxt (-compose #'lsp--severity-code->severity #'cl-first)
-                      (-compose #'length #'cl-rest)))
-         (-map (-lambda ((code count))
-                 (propertize (format "%s" count)
-                             'face (cl-case code
-                                     ('error 'error)
-                                     ('warning 'warning)
-                                     ('info 'success)
-                                     ('hint 'success)))))
-         (s-join "/")
-         (format "%s"))))
+                       ((equal :file lsp-diagnostics-modeline-scope)
+                        (list (lsp--get-buffer-diagnostics)))
+                       (t (->> (eq :workspace lsp-diagnostics-modeline-scope)
+                               (lsp-diagnostics)
+                               (ht-values))))))
+    (->> (let ((stats (make-vector lsp/diagnostic-severity-max 0))
+               strs
+               (i 0))
+           (mapc (lambda (buf-diags)
+                   (mapc (lambda (diag)
+                           (-let [(&Diagnostic? :severity?) diag]
+                             (when severity?
+                               (cl-incf (aref stats severity?)))))
+                         buf-diags))
+                 diagnostics)
+           (while (< i lsp/diagnostic-severity-max)
+             (when (> (aref stats i) 0)
+               (setq strs
+                     (nconc strs
+                            `(,(propertize
+                                (format "%s" (aref stats i))
+                                'face
+                                (cond
+                                  ((equal i lsp/diagnostic-severity-error) 'error)
+                                  ((equal i lsp/diagnostic-severity-warning) 'warning)
+                                  ((equal i lsp/diagnostic-severity-information) 'success)
+                                  ((equal i lsp/diagnostic-severity-hint) 'success)))))))
+             (cl-incf i))
+           strs)
+         (s-join "/"))))
+
+(defun lsp--diagnostics-reset-modeline-cache ()
+  ""
+  (plist-put lsp--diagnostics-modeline-wks->strings (car (lsp-workspaces)) nil)
+  (plist-put lsp--diagnostics-modeline-wks->strings :global nil)
+  (setq lsp--diagnostics-modeline-string nil))
```

Comment:
that would reset it only in the current buffer, right? 

(As a side note using only this variable for local caching should be enough, it will be fine if we recompute once per buffer).

Reply:
> that would reset it only in the current buffer, right?

Yes, but this is only need if the `modeline-scope` is `:file`. In that case only reset/update current buffer is totally fine.
For case of `:workspace` and `:global` the cached modeline is properly reset already.

Reply:
If you are in file A and update something that will result in a change of diagnostics in buffer B when you switch to B it will show stale diagnostics, right? I guess we might leave it as it is.

Reply:
> If you are in file A and update something that will result in a change of diagnostics in buffer B when you switch to B it will show stale diagnostics, right?

Ah, yes, it might be if the `modeline-scope` is `:file`. The default is `:workspace` so I guess it would be fine. We can try to reset all buffer strings but I doubt it might be a perf hit.